replit/poetry
# pip.use_pip_cache option to let pip fetch files

# Why?

Poetry is downloading .whl and .tar.gz files from pypi and then sending it to pip, which removes our custom pip's ability to get the files directly from the pip cacache. The downloading of large files can cause repls to run out of memory, thus failing to install large library like tensorflow.

## Change

1. add `POETRY_PIP_USE_PIP_CACHE` option. when on, poetry will ask pip to download files by passing URL instead of a file path
2. split _modern_install path from _traditional_install (pip) path to make logic easier to follow
3. add pip_install_from_url function to handle calling pip install with a URL
4. extracted _maybe_add_yanked_warning method for piecemeal reuse
5. fix accidentally broken test

## Testing

1. open [this repl](https://replit.com/join/lxpgayjbim-tobyho)
2. open `src/poetry/utils/env.py`, add `print('run_pip', cmd) after line 1481
3. `cd proj`
4. `python -m poetry add tabulate` and see that pip is called with a URL instead of file path
5. edit `.replit` setting `POETRY_PIP_USE_PIP_CACHE` to "0"
6. exit the shell
7. `cd proj`
8. `python -m poetry remove tabulate`
9. `python -m poetry add tabulate` and now see pip is called with a file path
10. cd `proj`
